### PR TITLE
Stale Synchronous Parallel Iterations

### PR DESCRIPTION
**Pull request by nltran**
Here is a pull request containing our development on [Stale Synchronous Parallel (SSP)](http://reports-archive.adm.cs.cmu.edu/anon/ml2013/CMU-ML-13-103.pdf) iterations. The pull request contains:
- Code supporting SSP
- API to configure and enable SSP from the ExecutionEnvironment

### CODE REVIEW

**Comment by StephanEwen** [comment]
Wow, impressive contribution! I would very much like to add SSP to Flink.

This is quite a lot, so it may take a bit to review and merge.

Here are some initial thoughts:

  - The SSP slack is defined as a job-wide parameter. Can you make it a parameter of the respective bulk iteration (that also has the mode PLAIN/SSP) ?

  - The code includes Apache Ignite as a parameter server. Is it possible to make this independent of Ignite? I know that SSP is usually used in parameter server ML settings, but I would like to not tie Ignite (another big dependency) into the Flink codebase.

**Comment by nltran** [comment]
>The code includes Apache Ignite as a parameter server. Is it possible to make this independent of Ignite? I know that SSP is usually used in parameter server ML settings, but I would like to not tie Ignite (another big dependency) into the Flink codebase.

I propose the following then:
1. Move the code related to the parameter server (model + Ignite implementation) currently in flink-core to a separate module.
2. Remove the parameter server startup code from the taskmanager.

This then implies that in order to use SSP and the parameter server the following needs to be done in the user job code:
1. Include the parameter server module as dependency to the project.
2. Start and stop the parameter server before and after each job.

How does that look for you?

**Comment by StephanEwen** [comment]
We are thinking about more general tooling for parameter servers (with implementations different from Ignite), because the request for that has come up. Whether we want to ship a parameter server with Flink, or simply add interfaces that help connecting against PS, is undecided, yet. 

For that reason, we should probably not overdesign this now. Let's add SSP to the runtime, week Ignite out, and make a nice self-contained example where Ignite is purely used in the user-code.

I think we can make a self-contained example in the following way:
The example MapFunction that communicates with the parameter server instantiates it's Ignite in the `open()` function, if it is in the first superstep. It then proceeds as in your code. 

To make this real-world applicable, I assume one would need dedicated Ignite processes anyways. In the example, the parameters are lost after the server is torn down, but so are they with an "in taskmanager parameter server", for all users of YARN as well (where TMs are shut down after a job).

The example here serves the purpose of demonstrating the capabilities and use of SSP, and to give people a blueprint how to connect this to a parameter server.

What do you think about this?

Her are some other comments:
  - Using Longs as parameter IDs is probably quite a bit cheaper than Strings.
  - The pull request also still contains markup from merging.
  - Some constants are hard-wired into the code. Would be good to pull them into static final fields at least, so it is very visible that these are predefined constants.
  - You seem to re-initialize the caches in every superstep. In general, initialization in `open()` in iterations needs to be done carefully, because `open()` is called at the beginning of each superstep. You can always check in which superstep you are using `getIterationRuntimeContext().getCurrentSuperstep()`

**Comment by nltran** [comment]
Thank you for your feedback!

I think this is a fine way to go. I will then provide the parameter server related code in a different module and include an example of how to run a SSP job with calls to the parameter server.

>You seem to re-initialize the caches in every superstep. In general, initialization in open() in iterations needs to be done carefully, because open() is called at the beginning of each superstep. You can always check in which superstep you are using getIterationRuntimeContext().getCurrentSuperstep()

We initialize the caches at the first superstep by checking if the reference is null, as the assumption is that since the same instantiated task is reused through the iterations, the cache is initialized only in the first superstep and the reference is not null in the subsequent supersteps. Would you recommend to do that check rather on the number of superstep?

**Comment by StephanEwen** [comment]
Ah, my bad. Concerning the initialization, I overlooked the null-check. It is good the way it is.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/base/BulkIterationStrategy.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2015 EURA NOVA.
```
Comment (rmetzger) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/base/BulkIterationStrategy.java]: I think you need to remove this copyright statement before we can merge the change.

File: flink-core/src/main/java/org/apache/flink/ps/impl/ParameterServerIgniteImpl.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+			this.ignite = Ignition.start(cfg1);
+
+			parameterCache = ignite.getOrCreateCache(parameterCacheCfg).withAsync();
+			sharedCache = ignite.getOrCreateCache(sharedCacheCfg).withAsync();
+
+			log.debug("I hereby confirm that parameter cache is async enabled: " + parameterCache.isAsync());
+			log.debug("I hereby confirm that shared cache is async enabled: " + sharedCache.isAsync());
+
+		} catch (Exception e) {
+			e.printStackTrace();
```
Comment (rmetzger) [inline flink-core/src/main/java/org/apache/flink/ps/impl/ParameterServerIgniteImpl.java]: Can you forward the exception as a runtime exception?

File: flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java [hunk 3]
```diff
@@ -920,15 +928,33 @@ private JobVertex createBulkIterationHead(BulkPartialSolutionPlanNode pspn) {
...
+			if(s.equals(BulkIterationStrategy.PLAIN) ) {
+				headVertex.setInvokableClass(IterationHeadPactTask.class);
+			}
+			else if(s.equals(BulkIterationStrategy.SSP)) {
+//			headVertex.setInvokableClass(ABSPIterationHeadPactTask.class);
+				headVertex.setInvokableClass(SSPIterationHeadPactTask.class);
+			}
+//=======
+//			headVertex.setInvokableClass(IterationHeadPactTask.class);
+//>>>>>>> upstream/master
```
Comment (rmetzger) [inline flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java]: merge conflict leftover?

Comment (nltran) [inline flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java]: Indeed, I am currently preparing a cleaner commit, sorry about that.

Comment (rmetzger) [inline flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java]: Okay ;)

**Comment by nltran** [comment]
The code now passes all the checks :smiley:  I have addressed the previous comments you made previously. @StephanEwen , @fhueske could you give it a another review? Namely:
* The SSP slack configuration is now in TaskConfig instead of job-wide. At some point we might want to unify both iterations strategies since the BSP iteration mode is an edge case for SSP with the slack equal to zero.

* The parameter server is now completely orthogonal to Flink core. It is up to the user to set it up and call it. I'm preparing a sample job that uses both SSP and calls to a parameter server.

* Correct license header. Will I have to fill in a Contributor License Agreement at some point?

**Comment by fhueske** [comment]
Hi @nltran,

thanks for updating the PR! 
I don't have time for a thorough review now, but I'll answer your last question.
ASF homepage states: *["The ASF desires that all contributors of ideas, code, or documentation to the Apache projects complete, sign, and submit [...] an Individual Contributor License Agreement."](http://www.apache.org/licenses)*

So, yes it would be very nice, if you would sign and submit an ICLA. In case, your employer assigned you to work on an Apache project, also an CCLA is necessary (see link above). 
Thank you very much, Fabian

**Comment by StephanEwen** [comment]
@nltran How would your use of the parameter server work together with what is proposed in #1003 ?

**Comment by nltran** [comment]
@StephanEwen 
Both interfaces for the parameter and the parameter server are quite similar. The things I see are the following:
- The parameter server code is tightly weaved to the core
- Lots of synchronization and registration still happen via sleeps

File: flink-runtime/src/test/java/org/apache/flink/runtime/iterative/event/EventWithAggregatorsTest.java [hunk 4]
```diff
@@ -123,6 +169,31 @@ private IterationEventWithAggregators pipeThroughSerialization(IterationEventWit
...
+			ClockTaskEvent newEvent = event.getClass().newInstance();
+			newEvent.read(new InputViewDataInputStreamWrapper(in));
+			in.close();
+
+			return newEvent;
+		} catch (Exception e) {
+			System.err.println(e.getMessage());
+			e.printStackTrace();
+			Assert.fail("Test threw an exception: " + e.getMessage());
+			return null;
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/iterative/event/EventWithAggregatorsTest.java]: `null` will never be returned. Why do you catch the exception here don't let it bubble up?

Comment (nltran) [inline flink-runtime/src/test/java/org/apache/flink/runtime/iterative/event/EventWithAggregatorsTest.java]: I actually reproduced the same as in ```testSerializationOfEventWithAggregateValues```. What you say makes sense in both cases though.

**Comment by sachingoel0101** [comment]
@nltran , I went through you code and if I understand correctly, the slack-related synchronization is done by the Runtime itself using events. I like that idea more than mine which involved synchronization by the Parameter Server itself which is, well, sort of messy.
Could you rebase this to the current master?

**Comment by nltran** [comment]
Hello guys,

Do you have any plans or updates on this one?

**Comment by hsaputra** [comment]
@nltran Looks like the current state of PR has conflict. Would you mind rebasing with latest?
Thx!

**Comment by StephanEwen** [comment]
I think the SSP code look in good shape.

Concerning the parameter server code, I think we should not add this this to Flink's repository, but it should be hosted in an external repository. We have to be a bit cautious at this point to not overload the Flink code base.

I would very much like to get a culture starting where people also build and offer libraries and tools for Flink in their own (3rd party) repositories. It has the advantage that teh main authors are in control of the code. Bugfixes can be applied fast, release can be made frequently (no Apache processes). The core repository stays slim, which makes it easier to maintain.

  - One possible place for the parameter server code would be a repository under https://github.com/project-flink . We can create one and grant you access rights.

  - We can link the parameter server code under http://flink.apache.org/community.html#third-party-packages

What we would need to do is split up this code into the SSP part (gets into the Flink code) and the Parameter Server part. The SSP part can probably be kept independent from any use with parameter servers (even though the most common and meaningful application is in conjunction with a parameter server).

Any thoughts on this?

**Comment by nltran** [comment]
An ecosystem of 3rd party libraries for Flink makes sense to me. That would be a good place for experimental features (à la Google Labs back in its time) that are somehow backed by the Flink community. That would be a good place for the parameter server code in its current state, as it not a finished product as the thrid-packages that are listed on  http://flink.apache.org/community.html#third-party-packages

The code for the parameter server is separated in https://github.com/apache/flink/pull/1102, with a generic interface.

**Comment by aljoscha** [comment]
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit.

**Pull request closed without merging**